### PR TITLE
Multi-node training support

### PR DESCRIPTION
**Pull request by michaelbenayoun**
This PR adds support for multi-node training.

- [ ] Check training TP + PP on multiple nodes with `llama-7b`
- [ ] Check training TP + PP on multiple nodes with `llama-70b`
- [ ] Check that files are pushed to the remote cache properly (with and without custom remote cache)
- [ ] Test evaluation with PP on multiple nodes
- [ ] Check that the logged loss and metrics are correct

In coming PRs:

- Add better support for gradient checkpointing (full vs feature to specify which layers to use)
- Handle GQA better with the custom layer for `neuronx_distributed`

### CODE REVIEW

File: optimum/neuron/distributed/decoder_models.py [hunk 1]
```diff
@@ -122,24 +122,43 @@ def _parallelize(
         device: Optional["torch.device"] = None,
         parallelize_embeddings: bool = True,
         sequence_parallel_enabled: bool = False,
+        should_parallelize_layer_predicate_func: Optional[Callable[["torch.nn.Module"], bool]] = None,
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/decoder_models.py:125]: same

File: optimum/neuron/distributed/decoder_models.py [hunk 2]
```diff
@@ -558,22 +606,44 @@ def _parallelize(
...
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/decoder_models.py:609]: same

File: optimum/neuron/distributed/decoder_models.py [hunk 3]
```diff
@@ -605,11 +675,22 @@ def transform(
         layer: "torch.nn.Module",
         sequence_parallel_enabled: bool = False,
         device: Optional["torch.device"] = None,
+        should_parallelize_layer_predicate_func: Optional[Callable[["torch.nn.Module"], bool]] = None,
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/decoder_models.py:678]: +1

File: optimum/neuron/distributed/decoder_models.py [hunk 4]
```diff
@@ -780,21 +862,44 @@ def _parallelize(
         device: Optional["torch.device"] = None,
         parallelize_embeddings: bool = True,
         sequence_parallel_enabled: bool = False,
+        should_parallelize_layer_predicate_func: Optional[Callable[["torch.nn.Module"], bool]] = None,
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/decoder_models.py:865]: +1

File: optimum/neuron/distributed/encoder_decoder_models.py [hunk 5]
```diff
@@ -54,9 +54,13 @@ def transform(
         cls,
         model: "PreTrainedModel",
         layer: "torch.nn.Module",
-        device: Optional["torch.device"] = None,
         sequence_parallel_enabled: bool = False,
+        device: Optional["torch.device"] = None,
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/encoder_decoder_models.py:58]: no need for quotation neither. And all following since torch is imported.

File: optimum/neuron/distributed/utils.py [hunk 6]
```diff
@@ -615,6 +670,34 @@ def parameter_can_be_initialized(model: torch.nn.Module, parent_module: torch.nn
     )
 
 
+@requires_neuronx_distributed
+def apply_activation_checkpointing(
+    model: Union[torch.nn.Module, "NxDPPModel"],
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/utils.py:675]: No need for quote neither

File: optimum/neuron/utils/training_utils.py [hunk 7]
```diff
@@ -367,7 +382,82 @@ def torch_xla_safe_save_file(
         save_file(cpu_data, filename, metadata=metadata)
 
 
-def get_model_param_count(model, trainable_only=False):
-    """Wrapper around `transformers.trainer_pt_utils.get_model_param_count` to handle tensor parallelism."""
-    # TODO: make it work for TP
-    return transformers_get_model_param_count(model, trainable_only=trainable_only)
+@requires_neuronx_distributed
+def get_model_param_count(model: Union[torch.nn.Module, "NxDPPModel"], trainable_only: bool = False):
+    """Counts the number of parameters of `model`."""
```
Comment (JingyaHuang) [inline optimum/neuron/utils/training_utils.py:387]:
```suggestion
    """Counts the number of parameters of the model."""
```

**Review by JingyaHuang — APPROVED** [review]
Multi node is huge, thanks for the work 🎊 . The PR looks good to me, just left some minor nits. Wonder if @michaelbenayoun if it's possible to test the least functionality with CIs or other paths(I know setting up for multiple nodes is tricky...

File: optimum/neuron/utils/hub_neuronx_cache.py [hunk 8]
```diff
@@ -328,14 +332,16 @@ def patch_neuron_cc_wrapper():
 
 
 @requires_torch_neuronx
-def synchronize_hub_cache(cache_repo_id: Optional[str] = None):
+def synchronize_hub_cache(cache_url: Optional[CacheUrl] = None, cache_repo_id: Optional[str] = None):
```
Comment (dacorvo) [inline optimum/neuron/utils/hub_neuronx_cache.py:335]: I would rather take a string here and convert it to a CacherUrl (what you are currently doing in the CLI).
This will avoid spreading the knowledge of the CacheUrl xla class to other modules.

File: optimum/commands/neuron/cache.py [hunk 9]
```diff
@@ -165,9 +167,21 @@ class SynchronizeRepoCommand(BaseOptimumCLICommand):
...
 
+    @requires_torch_neuronx
     def run(self):
-        synchronize_hub_cache(self.args.repo_id)
+        from libneuronxla.neuron_cc_cache import CacheUrl
+
+        if self.args.cache_dir is not None:
+            if not Path(self.args.cache_dir).is_dir():
+                raise ValueError(f"The {self.args.cache_dir} directory does not exist.")
+            cache_url = CacheUrl(self.args.cache_dir, url_type="fs")
```
Comment (dacorvo) [inline optimum/commands/neuron/cache.py:181]: I'd rather move this inside the synchronize_hub_cache method, to avoid importing CacheUrl here.

**Comment by michaelbenayoun** [comment]
We test settings with TP + PP, but testing the multi-node setting is more tricky, it requires to create a node cluster, etc. For now let's keep it like that.

**Review by dacorvo — COMMENTED** [review]
I only looked at the part that is related to the cache: I would rather move the import inside the utils to avoid multiple imports.

File: optimum/neuron/utils/hub_neuronx_cache.py [hunk 10]
```diff
@@ -332,15 +332,23 @@ def patch_neuron_cc_wrapper():
...
 @requires_torch_neuronx
-def synchronize_hub_cache(cache_url: Optional[CacheUrl] = None, cache_repo_id: Optional[str] = None):
+def synchronize_hub_cache(cache_path: Optional[Union[str, Path]] = None, cache_repo_id: Optional[str] = None):
     """Synchronize the neuronx compiler cache with the optimum-neuron hub cache.
 
     Args:
-        cache_url (`Optional[CacheUrl]`, defaults to `None`):
-            The cache url to use for synchronization.
+        cache_url (`Optional[Union[str, Path]]`, defaults to `None`):
+            The path of th to use for synchronization.
```
Comment (dacorvo) [inline optimum/neuron/utils/hub_neuronx_cache.py:340]:
```suggestion
            The path of the folder to use for synchronization.
```

**Review by dacorvo — APPROVED** [review]

**Comment by michaelbenayoun** [comment]
Merging, checked on custom instanced.

**Comment by shamanez** [comment]
@michaelbenayoun  can we conduct multi-node training with Mistral?

**Comment by michaelbenayoun** [comment]
Yes it should be!
We haven't run many experiments for now, but support is here.
